import { ReactComponent as CladeSchema } from 'src/assets/img/Nextstrain_ncov_clades-20B1tip.svg'

<!-- TODO: What is the main purpose of the tool? -->
<!-- TODO: How it is supposed to be used? -->
<!-- TODO: Who is the target audience? -->
<!-- TODO: Why we built the tool the way we did? -->
<!-- TODO: What are the trade-offs we have to choose? -->
<!-- TODO: What are the limitations? -->
<!-- TODO: What are the advantages of this tool compared to the existing alternatives? -->
<!-- TODO: How this tool complements the traditional workflow? -->

<!-- FIXME: what is "calls mutations"? We may want to explain this better -->

## What is Nextclade?

Nextclade is a free and open-source online tool that aligns your sequences to the reference sequence used by Nextstrain
(Wuhan-Hu-1 for SARS-CoV-2) and calls mutations relative to this reference. Clades are then assigned based on the
nucleotides at the positions used to
[define clades in Nextstrain](https://github.com/nextstrain/ncov/blob/master/config/clades.tsv). In addition, the tool
calculates a number of quality metrics. If these metrics exceed certain scores, the sequence will likely be excluded
from the Nextstrain builds.

## What are the clades?

<!-- FIXME: This section is specific to SARS-CoV-2, while all others try to be generic. This might be confising. -->

Since its emergence in late 2019, SARS-CoV-2 has diversified into several different co-circulating variants. To
facilitate discussion of these variants, we have grouped them into so called clades which are defined by specific
signature mutations.

We currently define 5 major clades (see [this blog post](https://nextstrain.org/blog/2020-06-02-SARSCoV2-clade-naming)
for details):

- 19A and 19B emerged in Wuhan and have been dominating the early outbreak
- 20A emerged from 19A out of dominated the European outbreak in March and has since spread globally
- 20B and 20C are large genetically distinct subclades 20A

<figure className="figure w-100 text-center">
  <picture className="w-75 figure-img" alt="illustration of the model">
    <CladeSchema height={200} />
  </picture>
  <figcaption>
    <small>Fig.1. Illustration of phylogenetic relationship of clades, as defined by nextstrain</small>
  </figcaption>
</figure>

## Processing steps

<!-- A positive-negative-positive sandwich for each section -->
<!-- Align with feature boxes -->

### Sequence Alignment

<!-- Sequence Alignment: In 1 phrase: remind user what it is (should be similar to the text in the corresp. feature box)  -->

<!-- Sequence Alignment: Give some specifics. Why is it important? Only positives. -->

<!-- Sequence Alignment: Discuss current limitations. Only negatives. -->

Nextclade does all computations on your computer -- your data does not leave your browser. This mean we can not perform
a full multiple sequence alignment or phylogenetic analysis. Instead, we roughly estimate the alignment of your
sequences to there reference one-by-one using short seeds (21-mers) along the sequence. With this rough overall
alignment, we perform a banded Smith-Waterman alignment with an affine gap-penalty.

If the seed matches are poor or suggest gaps in excess of 200 bases, we won't attempt to align -- your sequence won't be
scored.

<!-- Sequence Alignment: Finish with something positive  -->

For most sequences, however, this alignment strategy works well and takes less than a second per sequence.

### Clade Assignment

<!-- Clade Assignment: In 1 phrase: remind user what it is (should be similar to the text in the corresp. feature box)  -->

<!-- Clade Assignment: Give some specifics. Why is it important? Only positives. -->

<!-- Clade Assignment: Discuss current limitations. Only negatives. -->

The analysis on [nextstrain.org/ncov](https://nextstrain.org/ncov) uses the phylogenetic context to assign clades. Here,
we provide a simple tool to do this clade assignment sequence-by-sequence through alignment to the reference. This is
expected to be slightly less sensitive, but will still give the correct answer in most cases. Sometimes, however, clade
assignment will fail.

<!-- Clade Assignment: Finish with something positive  -->

### Quality Control (QC)

<!-- Quality Control: In 1 phrase: remind user what it is (should be similar to the text in the corresp. feature box)  -->

<!-- Quality Control: In 1 phrase: remind user what it is (should be similar to the text in the corresp. feature box)  -->

<!-- Quality Control: Give some specifics. Why is it important? Only positives. -->

<!-- Quality Control: Discuss current limitations. Only negatives. -->

The tool will scan your sequences for excess divergence and clusters of differences from the reference. These might
indicate problems during sequencing or bioinformatic assembly and would result in these sequences being filtered out by
Nextstrain.

We currently implemented four metrics to flag a sequence as potentially problematic:

- too much missing data: If your sequence misses more than 1000 sites (`N`s), it will be flagged
- too high divergence: Sequence more than 20 mutations away from the reference sequence are flagged
- too many ambiguous nucleotides: mixed states (such as `R`, `Y`, etc) are indicative of contamination (or
  superinfection) and more than 10 such non-ACGTN characters will result in a QC warning
- clustered differences: If your sequence has clusters with 6 or more differences in 100 bases (excluding known clusters
  like positions 28881-28883), it will be flagged.

These warnings don't necessarily mean your sequences are problematic, but these issues warrant closer examination. The
[Nextstrain pipeline](https://github.com/nextstrain/ncov) will likely exclude those sequences.

<!-- Clade Assignment: Finish with something positive  -->

## Features

<!-- A positive-negative-positive sandwich for each section -->
<!-- Align with feature boxes -->

### Fast iteration

### Ease of use

### Private

Importantly, these computations happen on your computer and ** your data does not leave your computer**

<!--
## How can I help?

Our project is free and open-source and we welcome contributions in any form.

We are glad to hear from you!
If you have encountered a problem or have an idea for an improvement: TODO
This will help us to identify problems and possibilities for improvements and to develop the project further.

If you want to participate in the discussion of the project feel free to join: TODO

Help us to translate the tool in your language: TODO

If you are a developer, take a look at:
 - our source code repository at githubL TODO
 - currently reported issues and feature requests: TODO
 - our developer's guide at: TODO
 -->

<!--
## Who are the authors?

 -->
