# About NextClades

**NextClades is currently under active development.
Implementation and formats are still fluid and the app likely contains bugs we don't know of.
Please report any issues and leave feedback at [github.com/neherlab/webclades](https://github.com/neherlab/webclades).**

NextClades aligns your sequences to the reference sequence used by Nextstrain (Wuhan-Hu-1 for SARS-CoV-2) and calls mutations relative to this reference.
Clades are then assigned based on the nucleotides at the positions used to [define clades in nextstrain](https://github.com/nextstrain/ncov/blob/master/config/clades.tsv).
In addition, the tool calculates a number of quality metrics.
If these metrics exceed certain scores, the sequence will likely be excluded from the nextstrain builds.

## Reference alignments
NextClades does all computations on your computer -- your data does not leave your browser.
This mean we can not perform a full multiple sequence alignment or phylogenetic analysis.
Instead, we roughly estimate the alignment of your sequences to there reference one-by-one using short seeds (21-mers) along the sequence.
With this rough overall alignment, we perform a banded Smith-Waterman alignment with an affine gap-penalty.

If the seed matches are poor or suggest gaps in excess of 200 bases, we won't attempt to align -- your sequence won't be scored.
For most sequences, however, this alignment strategy works well and takes less than a second per sequence.

## QC metrics
We currently implemented four different metrics to flag a sequence as potentially problematic:

 * too much missing data: If your sequence misses more than 1000 sites (`N`s), it will be flagged
 * too high divergence: Sequence more than 20 mutations away from the reference sequence are flagged
 * too many ambiguous nucleotides: mixed states (such as `R`, `Y`, etc) are indicative of contamination (or superinfection) and more than 10 such non-ACGTN characters will result in a QC warning
 * clustered differences: If your sequence has clusters with 6 or more differences in 100 bases (excluding known clusters like positions 28881-28883), it will be flagged.

These warnings don't necessarily mean your sequences are problematic, but these issues warrant closer examination.
The [Nextstrain pipeline](https://github.com/nextstrain/ncov) will likely exclude those sequences.

