---
title: 'Nextclade: clade assignment, mutation calling and QC for viral genomes'
tags:
  - phylogenetics
  - typing
  - sequence alignment
authors:
  - name: Ivan Aksamentov
    affiliation: "1, 2" # (Multiple affiliations must be quoted)
  - name: Emma B. Hodcroft
    affiliation: "2, 3"
  - name: James Hadfield
    affiliation: "4"
  - name: Trevor Bedford
    affiliation: "4, 5"
  - name: Richard A. Neher^[corresponding author]
    affiliation: "1, 2"
    orcid: 0000-0003-2525-1407
affiliations:
 - name: Biozentrum, University of Basel, Switzerland
   index: 1
 - name: Swiss Institute of Bioinformatics, Basel, Switzerland
   index: 2
 - name: Institute of Social and Preventive Medicine, University of Bern, Bern, Switzerland
   index: 3
 - name: Molecular and Cellular Biology Program, University of Washington, Seattle, WA, USA
   index: 4
 - name: Vaccine and Infectious Disease Division, Fred Hutchinson Cancer Research Center, Seattle, WA, USA
   index: 5
date: 15 July 2021
bibliography: paper.bib

---

# Summary



# Statement of need

After assembly of a consensus genome from read data, one usually wants to (i) assess the quality of the sequence, (ii) assign to a known clade or type, and (ii) compare it to a reference sequence to detect important evolutionary changes.
`Nextclade` addresses this need through via a command-line interface for bulk analysis of many sequences and a web-tool with the same functionality.
`Nextclade` is build on `Nextalign`, a codon-aware pairwise sequence aligner for similar viral genomes, which allows unabiguous calling of amino acid changes associated with changes in the nucleotide sequence.
The sequence is then placed onto a phylogenetic tree generated by the `augur` pipeline [@huddleston_augur_2021] of the `Nextstrain` tool-chain [@hadfield_nextstrain_2018].

During the SARS-CoV-2 pandemic, `Nextclade` has already allowed countless users to quickly analyze their data, assign sequences to clades and variants of concern, and identify important mutations.

# Implementation
`Nextclade` consists of the packages `nextalign`, `nextalign_cli`, `nextclade`, `nextclade_cli`, and the `web` tool.
The first four are implemented in C++ and represent the libary and CLI components of the two tools `nextalign` and `nextclade`.
The web-tool uses web assembly to compile the C++ code and interfaces it with a React application written in type-script.
All tools are meant to align several (many) sequences to one common reference sequence.

## nextalign

Nextalign implementes a banded pairwise Smith-Waterman alignment with an affine gap cost.
The bandwidth and relative shift of the two sequences are determined by simply seed matching.
In contrast to most other existing tools (e.g.~`minimap2` [@li_minimap2_2018] or `mafft` [@katoh_mafft_2013]), `nextalign` can ingest a genome annotation specifying codon regions and will then make the gap-opening penalty dependent on the reading frame.
This allows `nextalign` to choose the most biologically interpretable between otherwise equivalent alignments.
In the following example, the gap could be moved forward or backward by one base with the same number of matches, but a frame-dependent gap-opening penalty locks the gap in-frame:
```
...GTT.TAT.TAC...
...GTT.---.TAC...
```
Similarly, nextalign preferentially places gaps outside of genes in case of ambiguities.

In addition to nucleotide alignments, `nextalign` will extract the aligned coding sequences, translates them, and perform pairwise amino acid alignments.
These amino acid alignments are output along side the nucleotide alignment and are used by `nextclade` to determine amino acid changes.
All alignment parameters can be configured via CLI flags.

## nextclade

`Nextclade` uses `nextalign` to determine all mutations of each sequence relative to the reference sequence.
With this set of mutations, it performs an exhaustive search for the closest match on a phylogenetic tree of representing the diversity of the population, and deduces a clade annotation from the placement on the phylogeny.
In addition, it determines the mutations separating the closest match from the query sequence.
This set of ``private mutations'' are used as a QC metric: many private mutations are often a sign of sequencing errors or miscalled bases.
If such private mutations cluster in shprt stretches on the genome, this is an additional sign of concern.
The private mutation count, a measure of SNP clusters, as well as rules quantifying sequence completeness, characters indicating ambiguous bases, stop-codons, and frame-shifts are used to quantify sequence quality individually and via an aggregate score.


## web interface

While CLI tools are most appropriate for bulk processing, processing of up to a few hundreds of sequences is feasible and possibly more convenient via a graphical interface coupled to a visualization.
Nextclade enables this via a completely client side web-application onto which users can drop a fasta files with sequences.
The results are displayed in an interactive viewer that highlights QC results nucleotide mutations (see \autoref{overview}) and allows to explore the effects of complex mutations on the viral proteins (see \autoref{fig:tooltip}).


![Overview of the QC results and nucleotide mutation view.\label{fig:screenshot_overview}](figures/overview.png)

![Mutations in each gene can be explored interactively and show how the changes in the nucleotide sequence correspond to changes in the viral protein.\label{fig:tooltip}](figures/tooltip.png)


# Discussion and Outlook

Nextclade was implemented to respond to the increasing need for laboratories around the world to quickly assess the quality of their newly generated data, categorize them into different variants and clades, and investigate their mutational profiles.
While Nextclade has some similarities to UShER [@turakhia_ultrafast_2021], these two tools address different use cases.
UShER allows to place sequences on a comprehensive tree of 100s of thousands of leaves and further refines the phylogenetic relationship of the user supplied sequences to analyze the fine-scale relationship of the user supplied sequences and other publicly available data.
Nextclade provides a completely client side analysis of sequences with a focus on QC, clade assignment, and investigation of variation.

Nextalign was written for a very specific use case: pairwise alignment of similar sequences ($<10%$ divergence) with limited insertions and deletions.
For more general use cases, tools like `mafft` or `minimap2` are likely more robust and further improvements to the alignment algorithms are possible.

Going forward, we plan to extend Nexclade to other viruses including the different seasonal influenza viruses.


# Acknowledgements

We gratefully acknowledge the generous publicly sharing of sequence data by many labs around the world that make tools like Nextclade useful.
We are also grateful for feedback from the Nextstrain team and the wider community for critical feedback and suggestions on how to improve the tools.

# References
